{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "911ae6e6-9eb8-4e6e-90d2-7265f726e06e",
   "metadata": {},
   "source": [
    "# NOTE:\n",
    "This is the same tutorial notebook from the main UCE repo. We have re run it with tabula sapiens v2 and the human brain cell atlas, which take a long time to run, for reproducibility and to demonstrate how to run the benchmark for large datasets by repeatedly resampling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b258384-9a56-4ed0-be6f-db1c94711356",
   "metadata": {},
   "source": [
    "# Large Scale Embedding benchmarks\n",
    "\n",
    "This notebook includes an example showing how to run large scale embedding benchmarks using scIB [(single-cell integration benchmark)](https://www.nature.com/articles/s41592-021-01336-8)\n",
    "\n",
    "We use the GPU accelerated version implemented here: https://github.com/YosefLab/scib-metrics\n",
    "\n",
    "Please follow installation instructions in that repo. \n",
    "\n",
    "*Note: installing Faiss can be difficult and may take some time*\n",
    "\n",
    "*Running the full benchmarking suite on many cells can take many hours, even on GPUs with large amounts of memory, such as A100s, and with many threads*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca4ba3a1-5c85-4c7b-8564-f8c5689e9345",
   "metadata": {},
   "source": [
    "## Load Imports and define Benchmark Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b9d9fd58-915b-492d-9880-48c37e3859a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/scanpy/_utils/__init__.py:35: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.\n",
      "  from anndata import __version__ as anndata_version\n",
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/anndata/__init__.py:70: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  return module_get_attr_redirect(attr_name, deprecated_mapping=_DEPRECATED)\n",
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/anndata/__init__.py:70: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  return module_get_attr_redirect(attr_name, deprecated_mapping=_DEPRECATED)\n",
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/anndata/__init__.py:70: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  return module_get_attr_redirect(attr_name, deprecated_mapping=_DEPRECATED)\n",
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/anndata/__init__.py:70: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  return module_get_attr_redirect(attr_name, deprecated_mapping=_DEPRECATED)\n",
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/anndata/__init__.py:70: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  return module_get_attr_redirect(attr_name, deprecated_mapping=_DEPRECATED)\n",
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/anndata/__init__.py:70: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  return module_get_attr_redirect(attr_name, deprecated_mapping=_DEPRECATED)\n",
      "/lfs/ampere5/0/yanay/env/micromamba/envs/dogma/envs/faiss_1.8.0/lib/python3.11/site-packages/anndata/__init__.py:70: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  return module_get_attr_redirect(attr_name, deprecated_mapping=_DEPRECATED)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc\n",
    "\n",
    "from scib_metrics.benchmark import Benchmarker\n",
    "\n",
    "import faiss\n",
    "\n",
    "from scib_metrics.nearest_neighbors import NeighborsResults\n",
    "\n",
    "# Faiss GPU accelerate nearest neighbors methods\n",
    "def faiss_hnsw_nn(X: np.ndarray, k: int):\n",
    "    \"\"\"Gpu HNSW nearest neighbor search using faiss.\n",
    "\n",
    "    See https://github.com/nmslib/hnswlib/blob/master/ALGO_PARAMS.md\n",
    "    for index param details.\n",
    "    \"\"\"\n",
    "    X = np.ascontiguousarray(X, dtype=np.float32)\n",
    "    res = faiss.StandardGpuResources()\n",
    "    M = 32\n",
    "    index = faiss.IndexHNSWFlat(X.shape[1], M, faiss.METRIC_L2)\n",
    "    gpu_index = faiss.index_cpu_to_gpu(res, 0, index)\n",
    "    gpu_index.add(X)\n",
    "    distances, indices = gpu_index.search(X, k)\n",
    "    del index\n",
    "    del gpu_index\n",
    "    # distances are squared\n",
    "    return NeighborsResults(indices=indices, distances=np.sqrt(distances))\n",
    "\n",
    "\n",
    "def faiss_brute_force_nn(X: np.ndarray, k: int):\n",
    "    \"\"\"Gpu brute force nearest neighbor search using faiss.\"\"\"\n",
    "    X = np.ascontiguousarray(X, dtype=np.float32)\n",
    "    res = faiss.StandardGpuResources()\n",
    "    index = faiss.IndexFlatL2(X.shape[1])\n",
    "    gpu_index = faiss.index_cpu_to_gpu(res, 0, index)\n",
    "    gpu_index.add(X)\n",
    "    distances, indices = gpu_index.search(X, k)\n",
    "    del index\n",
    "    del gpu_index\n",
    "    # distances are squared\n",
    "    return NeighborsResults(indices=indices, distances=np.sqrt(distances))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4c5fb90f-ffa5-4cb9-bf6a-6afce956fc86",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from scib_metrics.benchmark import Benchmarker, BioConservation, BatchCorrection\n",
    "import pandas as pd\n",
    "\n",
    "## Benchmarking Function, returns dataframe of scores\n",
    "def benchmark(ad, label_key=\"cell_type\", batch_key=\"sample_id\", obsm_keys=[\"X_uce\", \"X_scGPT\", \"X_geneformer\"], do_batchcons=False):\n",
    "    print(f\"Running using CT key:\", label_key)\n",
    "    biocons = BioConservation()\n",
    "    if do_batchcons:\n",
    "        batchcons = BatchCorrection(pcr_comparison=False)\n",
    "    else:\n",
    "        batchcons = BatchCorrection(ilisi_knn=False, kbet_per_label=False, graph_connectivity=False, pcr_comparison=False)\n",
    "    bm = Benchmarker(\n",
    "        ad,\n",
    "        batch_key=batch_key,\n",
    "        label_key=label_key,\n",
    "        embedding_obsm_keys=obsm_keys,\n",
    "        bio_conservation_metrics=biocons,\n",
    "        batch_correction_metrics=batchcons,\n",
    "        n_jobs=64,\n",
    "    )\n",
    "    bm.prepare(neighbor_computer=faiss_brute_force_nn)\n",
    "    bm.benchmark()\n",
    "    df = bm.get_results(min_max_scale=False)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f3bb257-21d4-41d5-9726-50b5e7af04b2",
   "metadata": {},
   "source": [
    "### Load in anndata\n",
    "\n",
    "For this example, we will benchmark cells from developing mouse brain.\n",
    "\n",
    "You can download an anndata object with UCE, scGPT and Geneformer embeddings precalulated from [here](https://drive.google.com/drive/folders/1f63fh0ykgEhCrkd_EVvIootBw7LYDVI7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed9ee327-7587-4b5c-83aa-9c3ddbc4fbdc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 581430 × 45792\n",
       "    obs: 'donor', 'tissue', 'anatomical_position', 'method', 'cdna_plate', 'library_plate', 'notes', 'cdna_well', 'old_index', 'assay', 'sample_id', 'sample', 'replicate', '10X_run', '10X_barcode', 'ambient_removal', 'donor_method', 'donor_assay', 'donor_tissue', 'donor_tissue_assay', 'cell_ontology_class', 'cell_ontology_id', 'compartment', 'broad_cell_class', 'free_annotation', 'manually_annotated', 'published_2022', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ercc', 'pct_counts_ercc', '_scvi_batch', '_scvi_labels', 'scvi_leiden_donorassay_full', 'age', 'sex', 'ethnicity', 'n_genes', 'donor_num', 'cell_type_coarse', 'n_counts'\n",
       "    var: 'n_cells'\n",
       "    uns: 'log1p', 'neighbors', 'umap'\n",
       "    obsm: 'X_geneformer', 'X_scarches', 'X_scgpt', 'X_scvi', 'X_tgpt', 'X_uce'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tabula_ad = sc.read(\"export_data/new_tabula_scib.h5ad\")\n",
    "tabula_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4cb1a5e-1672-4ba7-b488-036de0e3ff61",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_column = \"cell_ontology_class\"\n",
    "batch_column = \"sample_id\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "134f4e09-8e68-43fb-9d12-d87a1b5318c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tabula_ad.obs[cell_type_column].unique()) # Number of unique cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ac956e69-9a66-4225-adb8-a01a2d6e23bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "167"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tabula_ad.obs[batch_column].unique()) # Number of unique batches"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c790b976-1e84-40e2-ac07-8c12239e029c",
   "metadata": {},
   "source": [
    "# Running the Benchmark on the full dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8fedf0ac-7ea6-427d-b94f-1599c973fb02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# X_geneformer', 'X_scgpt', 'X_uce', 'X_tgpt', 'X_scvi', 'X_scarches'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87f63bba-8ae3-461e-bb32-6dcf95a6b7dd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running using CT key: cell_ontology_class\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing neighbors:   0%|                                | 0/6 [00:00<?, ?it/s]"
     ]
    }
   ],
   "source": [
    "new_tabula_benchmark_results_df = benchmark(tabula_ad, label_key=cell_type_column,  batch_key=batch_column, obsm_keys=['X_uce', 'X_geneformer',  'X_scgpt', \"X_tgpt\", 'X_scvi', 'X_scarches'], do_batchcons=True)\n",
    "new_tabula_benchmark_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "52b26e77-a73b-4d9d-a6ea-43171ba85833",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: For scVI and scARCHES, the results are just for one seed (seed 0) shown here\n",
    "# The result in Supplementary Table 1 is after running 10 seeds for each and averaging scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e7b2f258-1212-46ba-8d8a-9f88ecc096c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolated labels</th>\n",
       "      <th>KMeans NMI</th>\n",
       "      <th>KMeans ARI</th>\n",
       "      <th>Silhouette label</th>\n",
       "      <th>cLISI</th>\n",
       "      <th>Silhouette batch</th>\n",
       "      <th>iLISI</th>\n",
       "      <th>KBET</th>\n",
       "      <th>Graph connectivity</th>\n",
       "      <th>Batch correction</th>\n",
       "      <th>Bio conservation</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Embedding</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X_uce</th>\n",
       "      <td>0.647063</td>\n",
       "      <td>0.704803</td>\n",
       "      <td>0.229917</td>\n",
       "      <td>0.537544</td>\n",
       "      <td>0.999913</td>\n",
       "      <td>0.87662</td>\n",
       "      <td>0.006863</td>\n",
       "      <td>0.414848</td>\n",
       "      <td>0.691408</td>\n",
       "      <td>0.497434</td>\n",
       "      <td>0.623848</td>\n",
       "      <td>0.573283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_geneformer</th>\n",
       "      <td>0.530591</td>\n",
       "      <td>0.553988</td>\n",
       "      <td>0.142943</td>\n",
       "      <td>0.454141</td>\n",
       "      <td>0.99937</td>\n",
       "      <td>0.846715</td>\n",
       "      <td>0.006589</td>\n",
       "      <td>0.312182</td>\n",
       "      <td>0.644451</td>\n",
       "      <td>0.452484</td>\n",
       "      <td>0.536207</td>\n",
       "      <td>0.502718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_scgpt</th>\n",
       "      <td>0.452378</td>\n",
       "      <td>0.290269</td>\n",
       "      <td>0.039445</td>\n",
       "      <td>0.403705</td>\n",
       "      <td>0.993694</td>\n",
       "      <td>0.816081</td>\n",
       "      <td>0.036039</td>\n",
       "      <td>0.29483</td>\n",
       "      <td>0.349412</td>\n",
       "      <td>0.37409</td>\n",
       "      <td>0.435898</td>\n",
       "      <td>0.411175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_tgpt</th>\n",
       "      <td>0.555026</td>\n",
       "      <td>0.497</td>\n",
       "      <td>0.111297</td>\n",
       "      <td>0.449734</td>\n",
       "      <td>0.998918</td>\n",
       "      <td>0.875747</td>\n",
       "      <td>0.007018</td>\n",
       "      <td>0.298695</td>\n",
       "      <td>0.598287</td>\n",
       "      <td>0.444937</td>\n",
       "      <td>0.522395</td>\n",
       "      <td>0.491412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_scvi</th>\n",
       "      <td>0.592119</td>\n",
       "      <td>0.670051</td>\n",
       "      <td>0.194243</td>\n",
       "      <td>0.511048</td>\n",
       "      <td>0.999397</td>\n",
       "      <td>0.822333</td>\n",
       "      <td>0.015383</td>\n",
       "      <td>0.435243</td>\n",
       "      <td>0.701316</td>\n",
       "      <td>0.493569</td>\n",
       "      <td>0.593372</td>\n",
       "      <td>0.55345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_scarches</th>\n",
       "      <td>0.615001</td>\n",
       "      <td>0.681848</td>\n",
       "      <td>0.218949</td>\n",
       "      <td>0.542676</td>\n",
       "      <td>0.999307</td>\n",
       "      <td>0.827007</td>\n",
       "      <td>0.017227</td>\n",
       "      <td>0.42448</td>\n",
       "      <td>0.739538</td>\n",
       "      <td>0.502063</td>\n",
       "      <td>0.611556</td>\n",
       "      <td>0.567759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Metric Type</th>\n",
       "      <td>Bio conservation</td>\n",
       "      <td>Bio conservation</td>\n",
       "      <td>Bio conservation</td>\n",
       "      <td>Bio conservation</td>\n",
       "      <td>Bio conservation</td>\n",
       "      <td>Batch correction</td>\n",
       "      <td>Batch correction</td>\n",
       "      <td>Batch correction</td>\n",
       "      <td>Batch correction</td>\n",
       "      <td>Aggregate score</td>\n",
       "      <td>Aggregate score</td>\n",
       "      <td>Aggregate score</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Isolated labels        KMeans NMI        KMeans ARI  \\\n",
       "Embedding                                                            \n",
       "X_uce                 0.647063          0.704803          0.229917   \n",
       "X_geneformer          0.530591          0.553988          0.142943   \n",
       "X_scgpt               0.452378          0.290269          0.039445   \n",
       "X_tgpt                0.555026             0.497          0.111297   \n",
       "X_scvi                0.592119          0.670051          0.194243   \n",
       "X_scarches            0.615001          0.681848          0.218949   \n",
       "Metric Type   Bio conservation  Bio conservation  Bio conservation   \n",
       "\n",
       "              Silhouette label             cLISI  Silhouette batch  \\\n",
       "Embedding                                                            \n",
       "X_uce                 0.537544          0.999913           0.87662   \n",
       "X_geneformer          0.454141           0.99937          0.846715   \n",
       "X_scgpt               0.403705          0.993694          0.816081   \n",
       "X_tgpt                0.449734          0.998918          0.875747   \n",
       "X_scvi                0.511048          0.999397          0.822333   \n",
       "X_scarches            0.542676          0.999307          0.827007   \n",
       "Metric Type   Bio conservation  Bio conservation  Batch correction   \n",
       "\n",
       "                         iLISI              KBET Graph connectivity  \\\n",
       "Embedding                                                             \n",
       "X_uce                 0.006863          0.414848           0.691408   \n",
       "X_geneformer          0.006589          0.312182           0.644451   \n",
       "X_scgpt               0.036039           0.29483           0.349412   \n",
       "X_tgpt                0.007018          0.298695           0.598287   \n",
       "X_scvi                0.015383          0.435243           0.701316   \n",
       "X_scarches            0.017227           0.42448           0.739538   \n",
       "Metric Type   Batch correction  Batch correction   Batch correction   \n",
       "\n",
       "             Batch correction Bio conservation            Total  \n",
       "Embedding                                                        \n",
       "X_uce                0.497434         0.623848         0.573283  \n",
       "X_geneformer         0.452484         0.536207         0.502718  \n",
       "X_scgpt               0.37409         0.435898         0.411175  \n",
       "X_tgpt               0.444937         0.522395         0.491412  \n",
       "X_scvi               0.493569         0.593372          0.55345  \n",
       "X_scarches           0.502063         0.611556         0.567759  \n",
       "Metric Type   Aggregate score  Aggregate score  Aggregate score  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_tabula_benchmark_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3faf6f17-bbf7-4662-9e18-c196b8aa1daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'X_pca', 'X_scvi_seed0', 'X_scarches_seed0'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee280476-4057-4051-b4f1-eb7ee0055e69",
   "metadata": {},
   "source": [
    "# Running the Benchmark using Resampling (Human Brain Cell Atlas)\n",
    "\n",
    "Running the benchmark on the full dataset can take a very long time. Instead, we can run on medium sized samples of cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0cae96b8-5be1-4ea5-a919-d16d2205d645",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_size = 500_000 # number of cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1b90f5a4-da40-418d-a7c5-ddbfffe82954",
   "metadata": {},
   "outputs": [],
   "source": [
    "N_RESAMPLES = 2 # actually is 10 in the paper but this takes a very long time to run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ebe33620-5bbb-45a8-bb7e-c849ccee6676",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2480956 × 17928\n",
       "    obs: 'ROIGroup', 'ROIGroupCoarse', 'ROIGroupFine', 'roi', 'organism_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'assay_ontology_term_id', 'sex_ontology_term_id', 'development_stage_ontology_term_id', 'donor_id', 'suspension_type', 'dissection', 'fraction_mitochondrial', 'fraction_unspliced', 'cell_cycle_score', 'total_genes', 'total_UMIs', 'sample_id', 'supercluster_term', 'cluster_id', 'subcluster_id', 'cell_type_ontology_term_id', 'tissue_ontology_term_id', 'is_primary_data', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'n_genes', 'n_counts'\n",
       "    var: 'Biotype', 'Chromosome', 'End', 'Gene', 'Start', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'ensembl_id', 'n_cells'\n",
       "    uns: 'batch_condition', 'citation', 'log1p', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_geneformer', 'X_scGPT', 'X_uce'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hbca_ad = sc.read(\"export_data/brain_atlas_uce_scgpt_geneformer.h5ad\")\n",
    "hbca_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b6c280bb-8e9e-45b6-97f2-d4790cbcef6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_column = \"cluster_id\"\n",
    "batch_key = \"sample_id\" # not used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bfe80843-05d8-4412-8a89-b1332548f340",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "382"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(hbca_ad.obs[cell_type_column].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "189ad01d-83c0-40e6-ab13-d16ed7eb0c88",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "85fd7aafcdc24dc6a5944a773486938f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running using CT key: cluster_id\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing neighbors:   0%|                                                                                                     | 0/3 [00:00<?, ?it/s]\u001b[A\n",
      "Computing neighbors:  33%|███████████████████████████████                                                              | 1/3 [01:22<02:44, 82.49s/it]\u001b[A\n",
      "Computing neighbors:  67%|██████████████████████████████████████████████████████████████                               | 2/3 [01:58<00:55, 55.40s/it]\u001b[A\n",
      "Computing neighbors: 100%|█████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [02:35<00:00, 51.76s/it]\u001b[A\n",
      "Embeddings:   0%|\u001b[32m                                                                                                              \u001b[0m| 0/3 [00:00<?, ?it/s]\u001b[0m\n",
      "Metrics:   0%|\u001b[34m                                                                                                                \u001b[0m| 0/10 [00:00<?, ?it/s]\u001b[0m\u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                             \u001b[0m| 0/10 [00:00<?, ?it/s, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m██████▌                                                          \u001b[0m| 1/10 [20:26<3:03:55, 1226.13s/it, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m█████                                              \u001b[0m| 1/10 [20:26<3:03:55, 1226.13s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m██████████▍                                         \u001b[0m| 2/10 [26:49<1:37:22, 730.31s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m█████████████                                                    \u001b[0m| 2/10 [26:49<1:37:22, 730.31s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m███████████████████▌                                             \u001b[0m| 3/10 [43:16<1:38:54, 847.79s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m█████████████████████▌                                                  \u001b[0m| 3/10 [43:16<1:38:54, 847.79s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m█████████████████████████████▌                                            \u001b[0m| 4/10 [43:18<51:21, 513.51s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m██████████████████████████▊                                        \u001b[0m| 4/10 [43:18<51:21, 513.51s/it, Batch correction: silhouette_batch]\u001b[0m\u001b[A\n",
      "Embeddings:  33%|\u001b[32m████████████████████████████████▋                                                                 \u001b[0m| 1/3 [47:02<1:34:05, 2822.54s/it]\u001b[0m\u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                                                                \u001b[0m| 0/10 [00:00<?, ?it/s]\u001b[0m\u001b[A\n",
      "                                                                                                                                                     \u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                             \u001b[0m| 0/10 [00:00<?, ?it/s, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m██████▊                                                             \u001b[0m| 1/10 [06:38<59:42, 398.01s/it, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m█████▍                                                \u001b[0m| 1/10 [06:38<59:42, 398.01s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m██████████▊                                           \u001b[0m| 2/10 [09:21<34:40, 260.10s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m█████████████▍                                                     \u001b[0m| 2/10 [09:21<34:40, 260.10s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m████████████████████                                               \u001b[0m| 3/10 [15:46<37:00, 317.20s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m██████████████████████▏                                                   \u001b[0m| 3/10 [15:46<37:00, 317.20s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m█████████████████████████████▌                                            \u001b[0m| 4/10 [15:47<19:13, 192.28s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m██████████████████████████▊                                        \u001b[0m| 4/10 [15:47<19:13, 192.28s/it, Batch correction: silhouette_batch]\u001b[0m\u001b[A\n",
      "Embeddings:  67%|\u001b[32m█████████████████████████████████████████████████████████████████▎                                \u001b[0m| 2/3 [1:05:13<30:03, 1803.76s/it]\u001b[0m\u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                                                                \u001b[0m| 0/10 [00:00<?, ?it/s]\u001b[0m\u001b[A\n",
      "                                                                                                                                                     \u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                             \u001b[0m| 0/10 [00:00<?, ?it/s, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m██████▊                                                             \u001b[0m| 1/10 [06:25<57:52, 385.82s/it, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m█████▍                                                \u001b[0m| 1/10 [06:25<57:52, 385.82s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m██████████▊                                           \u001b[0m| 2/10 [09:09<33:58, 254.86s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m█████████████▍                                                     \u001b[0m| 2/10 [09:09<33:58, 254.86s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m████████████████████                                               \u001b[0m| 3/10 [15:34<36:40, 314.30s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m██████████████████████▏                                                   \u001b[0m| 3/10 [15:34<36:40, 314.30s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m█████████████████████████████▌                                            \u001b[0m| 4/10 [15:34<19:03, 190.55s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m██████████████████████████▊                                        \u001b[0m| 4/10 [15:34<19:03, 190.55s/it, Batch correction: silhouette_batch]\u001b[0m\u001b[A\n",
      "Embeddings: 100%|\u001b[32m██████████████████████████████████████████████████████████████████████████████████████████████████\u001b[0m| 3/3 [1:23:04<00:00, 1661.65s/it]\u001b[0m\u001b[A\n",
      "\n",
      "                                                                                                                                                     \u001b[A"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 500000 × 17928\n",
       "    obs: 'ROIGroup', 'ROIGroupCoarse', 'ROIGroupFine', 'roi', 'organism_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'assay_ontology_term_id', 'sex_ontology_term_id', 'development_stage_ontology_term_id', 'donor_id', 'suspension_type', 'dissection', 'fraction_mitochondrial', 'fraction_unspliced', 'cell_cycle_score', 'total_genes', 'total_UMIs', 'sample_id', 'supercluster_term', 'cluster_id', 'subcluster_id', 'cell_type_ontology_term_id', 'tissue_ontology_term_id', 'is_primary_data', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'n_genes', 'n_counts'\n",
       "    var: 'Biotype', 'Chromosome', 'End', 'Gene', 'Start', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'ensembl_id', 'n_cells'\n",
       "    uns: 'batch_condition', 'citation', 'log1p', 'schema_reference', 'schema_version', 'title', 'pca'\n",
       "    obsm: 'X_geneformer', 'X_scGPT', 'X_uce', 'X_pca'\n",
       "    varm: 'PCs'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running using CT key: cluster_id\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing neighbors:   0%|                                                                                                     | 0/3 [00:00<?, ?it/s]\u001b[A\n",
      "Computing neighbors:  33%|███████████████████████████████                                                              | 1/3 [00:36<01:13, 36.80s/it]\u001b[A\n",
      "Computing neighbors:  67%|██████████████████████████████████████████████████████████████                               | 2/3 [00:53<00:24, 24.85s/it]\u001b[A\n",
      "Computing neighbors: 100%|█████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [01:09<00:00, 23.26s/it]\u001b[A\n",
      "Embeddings:   0%|\u001b[32m                                                                                                              \u001b[0m| 0/3 [00:00<?, ?it/s]\u001b[0m\n",
      "Metrics:   0%|\u001b[34m                                                                                                                \u001b[0m| 0/10 [00:00<?, ?it/s]\u001b[0m\u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                             \u001b[0m| 0/10 [00:00<?, ?it/s, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m██████▌                                                           \u001b[0m| 1/10 [16:26<2:27:54, 986.07s/it, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m█████▏                                              \u001b[0m| 1/10 [16:26<2:27:54, 986.07s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m██████████▍                                         \u001b[0m| 2/10 [22:49<1:24:14, 631.83s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m█████████████                                                    \u001b[0m| 2/10 [22:49<1:24:14, 631.83s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m███████████████████▌                                             \u001b[0m| 3/10 [42:45<1:43:44, 889.23s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m█████████████████████▌                                                  \u001b[0m| 3/10 [42:45<1:43:44, 889.23s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m█████████████████████████████▌                                            \u001b[0m| 4/10 [42:46<53:51, 538.56s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m██████████████████████████▊                                        \u001b[0m| 4/10 [42:46<53:51, 538.56s/it, Batch correction: silhouette_batch]\u001b[0m\u001b[A\n",
      "Embeddings:  33%|\u001b[32m████████████████████████████████▋                                                                 \u001b[0m| 1/3 [46:35<1:33:11, 2795.81s/it]\u001b[0m\u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                                                                \u001b[0m| 0/10 [00:00<?, ?it/s]\u001b[0m\u001b[A\n",
      "                                                                                                                                                     \u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                             \u001b[0m| 0/10 [00:00<?, ?it/s, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m██████▊                                                             \u001b[0m| 1/10 [06:26<57:54, 386.10s/it, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m█████▍                                                \u001b[0m| 1/10 [06:26<57:54, 386.10s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m██████████▊                                           \u001b[0m| 2/10 [09:09<34:01, 255.18s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m█████████████▍                                                     \u001b[0m| 2/10 [09:09<34:01, 255.18s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m████████████████████                                               \u001b[0m| 3/10 [15:34<36:40, 314.42s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m██████████████████████▏                                                   \u001b[0m| 3/10 [15:34<36:40, 314.42s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m█████████████████████████████▌                                            \u001b[0m| 4/10 [15:35<19:04, 190.82s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m██████████████████████████▊                                        \u001b[0m| 4/10 [15:35<19:04, 190.82s/it, Batch correction: silhouette_batch]\u001b[0m\u001b[A\n",
      "Embeddings:  67%|\u001b[32m█████████████████████████████████████████████████████████████████▎                                \u001b[0m| 2/3 [1:04:38<29:48, 1788.28s/it]\u001b[0m\u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                                                                \u001b[0m| 0/10 [00:00<?, ?it/s]\u001b[0m\u001b[A\n",
      "                                                                                                                                                     \u001b[A\n",
      "Metrics:   0%|\u001b[34m                                                                             \u001b[0m| 0/10 [00:00<?, ?it/s, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m██████▊                                                             \u001b[0m| 1/10 [06:25<57:53, 385.95s/it, Bio conservation: isolated_labels]\u001b[0m\u001b[A\n",
      "Metrics:  10%|\u001b[34m█████▍                                                \u001b[0m| 1/10 [06:25<57:53, 385.95s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m██████████▊                                           \u001b[0m| 2/10 [09:22<35:02, 262.86s/it, Bio conservation: nmi_ari_cluster_labels_kmeans]\u001b[0m\u001b[A\n",
      "Metrics:  20%|\u001b[34m█████████████▍                                                     \u001b[0m| 2/10 [09:22<35:02, 262.86s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m████████████████████                                               \u001b[0m| 3/10 [15:47<37:10, 318.63s/it, Bio conservation: silhouette_label]\u001b[0m\u001b[A\n",
      "Metrics:  30%|\u001b[34m██████████████████████▏                                                   \u001b[0m| 3/10 [15:47<37:10, 318.63s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m█████████████████████████████▌                                            \u001b[0m| 4/10 [15:48<19:19, 193.21s/it, Bio conservation: clisi_knn]\u001b[0m\u001b[A\n",
      "Metrics:  40%|\u001b[34m██████████████████████████▊                                        \u001b[0m| 4/10 [15:48<19:19, 193.21s/it, Batch correction: silhouette_batch]\u001b[0m\u001b[A\n",
      "Embeddings: 100%|\u001b[32m██████████████████████████████████████████████████████████████████████████████████████████████████\u001b[0m| 3/3 [1:22:42<00:00, 1654.05s/it]\u001b[0m\u001b[A\n",
      "\n",
      "                                                                                                                                                     \u001b[A"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 500000 × 17928\n",
       "    obs: 'ROIGroup', 'ROIGroupCoarse', 'ROIGroupFine', 'roi', 'organism_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'assay_ontology_term_id', 'sex_ontology_term_id', 'development_stage_ontology_term_id', 'donor_id', 'suspension_type', 'dissection', 'fraction_mitochondrial', 'fraction_unspliced', 'cell_cycle_score', 'total_genes', 'total_UMIs', 'sample_id', 'supercluster_term', 'cluster_id', 'subcluster_id', 'cell_type_ontology_term_id', 'tissue_ontology_term_id', 'is_primary_data', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'n_genes', 'n_counts'\n",
       "    var: 'Biotype', 'Chromosome', 'End', 'Gene', 'Start', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'ensembl_id', 'n_cells'\n",
       "    uns: 'batch_condition', 'citation', 'log1p', 'schema_reference', 'schema_version', 'title', 'pca'\n",
       "    obsm: 'X_geneformer', 'X_scGPT', 'X_uce', 'X_pca'\n",
       "    varm: 'PCs'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from tqdm.auto import tqdm\n",
    "sample_score_dfs = []\n",
    "\n",
    "for i in tqdm(range(N_RESAMPLES)):\n",
    "    # benchmark one sample\n",
    "    # sample is drawn with random state i\n",
    "    subsample_ad = sc.pp.subsample(hbca_ad, copy=True, n_obs=sample_size, random_state=i)\n",
    "    sample_df = benchmark(subsample_ad, label_key=cell_type_column,  batch_key=batch_column)\n",
    "    # show the results for this sample\n",
    "    display(subsample_ad)\n",
    "    # add it to the results for all samples\n",
    "    sample_score_dfs.append(sample_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "131fee3d-108d-4685-8afb-ec2863775b2f",
   "metadata": {},
   "source": [
    "### Final Scores\n",
    "\n",
    "We can aggregate the scores from all the samples, taking the mean value (and standard deviation of the score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4bd62bbd-f8ac-453f-b172-76cff96e5fc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_mean = pd.concat([df.drop(\"Metric Type\").reset_index() for df in sample_score_dfs]).groupby(\"Embedding\").agg(np.mean)\n",
    "# Note: we drop the \"Metric Type\" row since it contains strings which we can't take the mean of"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8f0216e0-7d47-4c8c-a094-c27642b8669b",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_std = pd.concat([df.drop(\"Metric Type\").reset_index() for df in sample_score_dfs]).groupby(\"Embedding\").agg(np.std)\n",
    "# Note: we drop the \"Metric Type\" row since it contains strings which we can't take the std of"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8db86f16-ad9b-486d-935b-c69084866f6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolated labels</th>\n",
       "      <th>KMeans NMI</th>\n",
       "      <th>KMeans ARI</th>\n",
       "      <th>Silhouette label</th>\n",
       "      <th>cLISI</th>\n",
       "      <th>Silhouette batch</th>\n",
       "      <th>Batch correction</th>\n",
       "      <th>Bio conservation</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Embedding</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X_geneformer</th>\n",
       "      <td>0.441117</td>\n",
       "      <td>0.282088</td>\n",
       "      <td>0.039184</td>\n",
       "      <td>0.393853</td>\n",
       "      <td>0.990547</td>\n",
       "      <td>0.758299</td>\n",
       "      <td>0.758299</td>\n",
       "      <td>0.429358</td>\n",
       "      <td>0.560934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_scGPT</th>\n",
       "      <td>0.56094</td>\n",
       "      <td>0.596169</td>\n",
       "      <td>0.141131</td>\n",
       "      <td>0.483329</td>\n",
       "      <td>0.997043</td>\n",
       "      <td>0.809811</td>\n",
       "      <td>0.809811</td>\n",
       "      <td>0.555722</td>\n",
       "      <td>0.657358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_uce</th>\n",
       "      <td>0.580308</td>\n",
       "      <td>0.702658</td>\n",
       "      <td>0.25047</td>\n",
       "      <td>0.50807</td>\n",
       "      <td>0.997679</td>\n",
       "      <td>0.771417</td>\n",
       "      <td>0.771417</td>\n",
       "      <td>0.607837</td>\n",
       "      <td>0.673269</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Isolated labels KMeans NMI KMeans ARI Silhouette label     cLISI  \\\n",
       "Embedding                                                                       \n",
       "X_geneformer        0.441117   0.282088   0.039184         0.393853  0.990547   \n",
       "X_scGPT              0.56094   0.596169   0.141131         0.483329  0.997043   \n",
       "X_uce               0.580308   0.702658    0.25047          0.50807  0.997679   \n",
       "\n",
       "             Silhouette batch Batch correction Bio conservation     Total  \n",
       "Embedding                                                                  \n",
       "X_geneformer         0.758299         0.758299         0.429358  0.560934  \n",
       "X_scGPT              0.809811         0.809811         0.555722  0.657358  \n",
       "X_uce                0.771417         0.771417         0.607837  0.673269  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3a71d4f4-1555-4cea-8f8f-332075e00de6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Embedding\n",
       "X_geneformer    0.429358\n",
       "X_scGPT         0.555722\n",
       "X_uce           0.607837\n",
       "Name: Bio conservation, dtype: object"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_mean[\"Bio conservation\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c741510b-83c7-474e-bfee-15f461f750b6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Faiss",
   "language": "python",
   "name": "faiss_1.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
